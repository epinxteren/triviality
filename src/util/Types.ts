
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

/**
 * Make all properties in T optional
 */
export type Optional<T> = {
  [P in keyof T]?: T[P];
};

/**
 * Make all properties in T optional
 */
export type NoDuplicates<T> = {
  [P in keyof T]?: never;
};
