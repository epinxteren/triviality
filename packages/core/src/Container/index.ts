import { ServiceTag, SF, TagServicePair } from '../ServiceFactory';
import { forEach } from 'ramda';
import { ImmutableContainer } from './ImmutableContainer';

export * from './ImmutableContainer';
export * from './MutableContainer';
export * from './MutableLockableContainer';

export const setServices = (setService: (tag: ServiceTag, sf: SF) => void) => (sfs: TagServicePair[]): void =>
  void forEach(([name, service]) => setService(name, service), sfs);

export const setNewServiceToContainer = <T extends ImmutableContainer>({ hasService, setService }: T) => (k: ServiceTag, sf: SF): T => {
  if (hasService(k)) {
    throw new Error(`Service ${k} already exists`);
  }
  return setService(k, sf);
};
