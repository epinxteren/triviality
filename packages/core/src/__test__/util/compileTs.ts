import * as fs from 'fs';
import * as path from 'path';
/* tslint:disable:no-var-requires */
const exec = require('child-process-promise').exec;

export async function compileTs(directory: string, ts: string): Promise<string> {
  const tmp = path.join(directory, 'tmp.ts');
  if (fs.existsSync(tmp)) {
    fs.unlinkSync(tmp);
  }
  fs.writeFileSync(tmp, ts);
  try {
    const result = await exec(`yarn ts-node --no-cache ${tmp}`);
    fs.unlinkSync(tmp);
    return result.stdout.toString();
  } catch (e) {
    fs.unlinkSync(tmp);
    return Promise.reject(e.stderr.toString());
  }
}

export async function declarationOfTs(directory: string, ts: string): Promise<string> {
  const tmp = path.join(directory, 'tmp.ts');
  if (fs.existsSync(tmp)) {
    fs.unlinkSync(tmp);
  }
  fs.writeFileSync(tmp, ts);
  try {
    await exec('yarn tsc');
    fs.unlinkSync(tmp);
    const featureFile = tmp
      .replace('src', 'build/src')
      .replace('.ts', '.d.ts');
    return fs.readFileSync(featureFile,
    ).toString();
  } catch (e) {
    fs.unlinkSync(tmp);
    return Promise.reject(e.toString());
  }
}
