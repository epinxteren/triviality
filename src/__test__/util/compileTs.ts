import * as fs from 'fs';
import * as path from 'path';
/* tslint:disable:no-var-requires */
const exec = require('child-process-promise').exec;

export async function compileTs(directory: string, ts: string): Promise<string> {
  const tmp = path.join(directory, 'tmp.ts');
  if (fs.existsSync(tmp)) {
    fs.unlinkSync(tmp);
  }
  fs.writeFileSync(tmp, ts);
  try {
    const result = await exec(`node_modules/.bin/ts-node ${tmp}`);
    fs.unlinkSync(tmp);
    return result.stdout.toString();
  } catch (e) {
    fs.unlinkSync(tmp);
    return Promise.reject(e.stderr.toString());
  }
}

export async function declarationOfTs(directory: string, ts: string): Promise<string> {
  const tmp = path.join(directory, 'tmp.ts');
  if (fs.existsSync(tmp)) {
    fs.unlinkSync(tmp);
  }
  fs.writeFileSync(tmp, ts);
  try {
    await exec('node_modules/.bin/tsc');
    fs.unlinkSync(tmp);
    const definitionFile = tmp
      .replace('src', 'build/src')
      .replace('.ts', '.d.ts');
    return fs.readFileSync(definitionFile,
    ).toString();
  } catch (e) {
    fs.unlinkSync(tmp);
    return Promise.reject(e.toString());
  }
}
