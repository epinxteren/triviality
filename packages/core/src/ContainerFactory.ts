import { FeatureFactory } from './FeatureFactory';
import { invokeFeatureFactories, invokeFeatureFactory } from './invokeFeatureFactory';
import { callSetupServices, SetupFeature, SetupFeatureServices } from './Feature';
import { ServiceFunctionReferenceContainer } from './Container';
import { KernelFeature, KernelServices } from './Feature/KernelFeature';

/**
 * Container factory.
 */
export class ContainerFactory<S> {
  public static create = (): ContainerFactory<SetupFeatureServices & KernelServices> => {
    const container = new ContainerFactory<SetupFeatureServices & KernelServices>();
    return container
      .add(SetupFeature);
  };

  public constructor(private featureFactories: FeatureFactory[] = []) {
  }

  /**
   * Merge functional service factory.
   */
  public add<Services>(f1: FeatureFactory<Services, S>): ContainerFactory<S & Services> {
    return new ContainerFactory([...this.featureFactories, f1] as any) as any;
  }

  public async build(): Promise<S> {
    const container = new ServiceFunctionReferenceContainer();
    const kernelFeature: any = KernelFeature(container);
    const features: FeatureFactory[] = [kernelFeature, ...this.featureFactories];
    invokeFeatureFactories({ container, invoke: invokeFeatureFactory })(features);
    const services: S & SetupFeatureServices & KernelServices = container.build();
    await callSetupServices(services.setupCallbacks);
    return services;
  }
}
