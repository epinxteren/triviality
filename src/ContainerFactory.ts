import {
  Module,
  ModuleConstructor,
  ModuleConstructor as MC,
  ModuleRegistries as MR,
  StrippedModule as SM,
  ModuleOptionalRegistries as MO,
} from './Module';
import { getAllPropertyNames } from './util/getAllPropertyNames';
import memorize from 'lodash.memoize';
import { ContainerError } from './ContainerError';

/**
 * Container factory.
 */
export class ContainerFactory<C /* Container */, R /* Registry */ > {

  public static create(): ContainerFactory<{}, {}> {
    const factory = new ContainerFactory<{}, {}>({} as {});
    return (factory as any);
  }

  private modules: Module[] = [];
  private isBuild = false;

  private constructor(private container: C) {
  }

  /**
   * Only add modules as second argument when there are circular dependencies between them.
   */
  public add<M1 extends MO<R>, NC = (C & SM<M1>), NR = (R & MR<M1>)>(m1: MC<M1, C>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, NC = (C & SM<M1> & SM<M2>), NR = (R & MR<M1> & MR<M2>)>(m1: MC<M1, C & SM<M2>>, m2: MC<M2, C & SM<M1>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3>), NR = (R & MR<M1> & MR<M2> & MR<M3>)>(m1: MC<M1, C & SM<M2> & SM<M3>>, m2: MC<M2, C & SM<M1> & SM<M3>>, m3: MC<M3, C & SM<M1> & SM<M2>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, M4 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3> & SM<M4>), NR = (R & MR<M1> & MR<M2> & MR<M3> & MR<M4>)>(m1: MC<M1, C & SM<M2> & SM<M3> & SM<M4>>, m2: MC<M2, C & SM<M1> & SM<M3> & SM<M4>>, m3: MC<M3, C & SM<M1> & SM<M2> & SM<M4>>, m4: MC<M4, C & SM<M1> & SM<M2> & SM<M3>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, M4 extends MO<R>, M5 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5>), NR = (R & MR<M1> & MR<M2> & MR<M3> & MR<M4> & MR<M5>)>(m1: MC<M1, C & SM<M2> & SM<M3> & SM<M4> & SM<M5>>, m2: MC<M2, C & SM<M1> & SM<M3> & SM<M4> & SM<M5>>, m3: MC<M3, C & SM<M1> & SM<M2> & SM<M4> & SM<M5>>, m4: MC<M4, C & SM<M1> & SM<M2> & SM<M3> & SM<M5>>, m5: MC<M5, C & SM<M1> & SM<M2> & SM<M3> & SM<M4>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, M4 extends MO<R>, M5 extends MO<R>, M6 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6>), NR = (R & MR<M1> & MR<M2> & MR<M3> & MR<M4> & MR<M5> & MR<M6>)>(m1: MC<M1, C & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6>>, m2: MC<M2, C & SM<M1> & SM<M3> & SM<M4> & SM<M5> & SM<M6>>, m3: MC<M3, C & SM<M1> & SM<M2> & SM<M4> & SM<M5> & SM<M6>>, m4: MC<M4, C & SM<M1> & SM<M2> & SM<M3> & SM<M5> & SM<M6>>, m5: MC<M5, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M6>>, m6: MC<M6, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, M4 extends MO<R>, M5 extends MO<R>, M6 extends MO<R>, M7 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7>), NR = (R & MR<M1> & MR<M2> & MR<M3> & MR<M4> & MR<M5> & MR<M6> & MR<M7>)>(m1: MC<M1, C & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7>>, m2: MC<M2, C & SM<M1> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7>>, m3: MC<M3, C & SM<M1> & SM<M2> & SM<M4> & SM<M5> & SM<M6> & SM<M7>>, m4: MC<M4, C & SM<M1> & SM<M2> & SM<M3> & SM<M5> & SM<M6> & SM<M7>>, m5: MC<M5, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M6> & SM<M7>>, m6: MC<M6, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M7>>, m7: MC<M7, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, M4 extends MO<R>, M5 extends MO<R>, M6 extends MO<R>, M7 extends MO<R>, M8 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8>), NR = (R & MR<M1> & MR<M2> & MR<M3> & MR<M4> & MR<M5> & MR<M6> & MR<M7> & MR<M8>)>(m1: MC<M1, C & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8>>, m2: MC<M2, C & SM<M1> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8>>, m3: MC<M3, C & SM<M1> & SM<M2> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8>>, m4: MC<M4, C & SM<M1> & SM<M2> & SM<M3> & SM<M5> & SM<M6> & SM<M7> & SM<M8>>, m5: MC<M5, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M6> & SM<M7> & SM<M8>>, m6: MC<M6, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M7> & SM<M8>>, m7: MC<M7, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M8>>, m8: MC<M8, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, M4 extends MO<R>, M5 extends MO<R>, M6 extends MO<R>, M7 extends MO<R>, M8 extends MO<R>, M9 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9>), NR = (R & MR<M1> & MR<M2> & MR<M3> & MR<M4> & MR<M5> & MR<M6> & MR<M7> & MR<M8> & MR<M9>)>(m1: MC<M1, C & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9>>, m2: MC<M2, C & SM<M1> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9>>, m3: MC<M3, C & SM<M1> & SM<M2> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9>>, m4: MC<M4, C & SM<M1> & SM<M2> & SM<M3> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9>>, m5: MC<M5, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M6> & SM<M7> & SM<M8> & SM<M9>>, m6: MC<M6, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M7> & SM<M8> & SM<M9>>, m7: MC<M7, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M8> & SM<M9>>, m8: MC<M8, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M9>>, m9: MC<M9, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8>>): ContainerFactory<NC, NR>;
  public add<M1 extends MO<R>, M2 extends MO<R>, M3 extends MO<R>, M4 extends MO<R>, M5 extends MO<R>, M6 extends MO<R>, M7 extends MO<R>, M8 extends MO<R>, M9 extends MO<R>, M10 extends MO<R>, NC = (C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9> & SM<M10>), NR = (R & MR<M1> & MR<M2> & MR<M3> & MR<M4> & MR<M5> & MR<M6> & MR<M7> & MR<M8> & MR<M9> & MR<M10>)>(m1: MC<M1, C & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9> & SM<M10>>, m2: MC<M2, C & SM<M1> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9> & SM<M10>>, m3: MC<M3, C & SM<M1> & SM<M2> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9> & SM<M10>>, m4: MC<M4, C & SM<M1> & SM<M2> & SM<M3> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9> & SM<M10>>, m5: MC<M5, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M6> & SM<M7> & SM<M8> & SM<M9> & SM<M10>>, m6: MC<M6, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M7> & SM<M8> & SM<M9> & SM<M10>>, m7: MC<M7, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M8> & SM<M9> & SM<M10>>, m8: MC<M8, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M9> & SM<M10>>, m9: MC<M9, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M10>>, m10: MC<M10, C & SM<M1> & SM<M2> & SM<M3> & SM<M4> & SM<M5> & SM<M6> & SM<M7> & SM<M8> & SM<M9>>): ContainerFactory<NC, NR>;
  public add(...moduleClasses: Array<ModuleConstructor<any, C>>): ContainerFactory<any, any> {
    if (this.isBuild) {
      throw ContainerError.containerAlreadyBuild();
    }
    for (const moduleClass of moduleClasses) {
      this.addModule(new moduleClass(this.container));
    }
    return this;
  }

  public async build(): Promise<C & R> {
    if (this.isBuild) {
      throw ContainerError.containerAlreadyBuild();
    }
    this.isBuild = true;
    this.combineRegistries();
    this.setup();
    return this.container as any;
  }

  private combineRegistries() {
    const combined: { [name: string]: any[] } = {};
    for (const module of this.modules) {
      if (module.registries) {
        const registers: any = module.registries();
        for (const register of Object.getOwnPropertyNames(registers)) {
          if (!combined[register]) {
            combined[register] = [];
          }
          const registered: any[] = registers[register]();
          if (registered) {
            if (!registered.forEach) {
              throw ContainerError.wrongRegisterReturnType(register);
            }
            registered.forEach((value) => {
              combined[register].push(value);
            });
          }
        }
      }
    }
    for (const register of Object.getOwnPropertyNames(combined)) {
      this.defineContainerProperty(register, () => combined[register]);
    }
    this.defineContainerProperty('registries', () => this.container);
  }

  private addModule(module: Module) {
    this.modules.push(module);
    for (const name of getAllPropertyNames(module)) {
      const value = (module as any)[name];
      if (['registries', 'setup'].indexOf(name) >= 0) {
        continue;
      }
      if (typeof value === 'function') {
        this.defineContainerProperty(name, memorize(value, (...args) => JSON.stringify(args)));
        (module as any)[name] = (this.container as any)[name].bind(this.container);
      } else {
        this.defineContainerProperty(name, value);
      }
    }
  }

  private defineContainerProperty(name: string, value: any) {
    if ((this.container as any)[name]) {
      // Ignore if it's a reference to the container
      if (value === this.container) {
        return;
      }
      throw ContainerError.propertyOrServiceAlreadyDefined(name);
    }
    Object.defineProperty(this.container, name, { get: () => value, set: () => { throw ContainerError.containerIsLocked(name); } });
  }

  private setup() {
    for (const module of this.modules) {
      if (module.setup) {
        module.setup();
      }
    }
  }
}
