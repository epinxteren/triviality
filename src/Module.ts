import { Registries } from './Registry';
import { NoDuplicates, Omit, Optional } from './util/Types';

export interface Module<R extends Registries = {}> {

  toString(): string;

  setup?(): void;

  registries?(): R;
}

export type ModuleOptionalRegistries<R> = Module<Optional<R>>;

export type ModuleConstructor<Instance, C> = new (container: C) => Instance & NoDuplicates<C>;

export type StrippedModule<T extends Module> = T extends { container: any } ? Omit<T, 'container' | 'registries' | 'setup'> : Omit<T, 'registries' | 'setup'>;

export type ModuleRegistries<T extends Module<R>, R = {}> = T extends { registries(): R; } ? ReturnType<NonNullable<T['registries']>> : {};

export type ModuleDependency<T extends Module<R>, R = {}> = ModuleRegistries<T, R> & StrippedModule<T>;
