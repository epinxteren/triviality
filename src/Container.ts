import { ModuleDependency as C } from './Module';

export type Container<M1, M2 = null> = Readonly<M2 extends null | undefined ? C<M1> : (C<M1> & C<M2>)>;
