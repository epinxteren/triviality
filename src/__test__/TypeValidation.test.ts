import { compileTs, declarationOfTs } from './util/compileTs';

describe('TypeValidation', async () => {

  describe('ContainerFactory', () => {

    it('Container should not have dependencies by default', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';

        ContainerFactory
          .create()
          .build()
          .then((container) => {
            container.someNonExistingService();
          });
      `))
        .rejects
        .toMatch('Property \'someNonExistingService\' does not exist on type \'{}\'');
    });

    it('Should force the requirements of a Module', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';
        import { Container } from '../Container';

        class Module1 implements Module {
          public userService() {
            return 'John';
          }
        }

        class Module2 implements Module {
          public constructor(private container: Container<Module1>) {

          }

          public halloService() {
            const userService = this.container.userService();
            return { speak: () => 'Hallo ' + userService }
          }
        }

        export const container = ContainerFactory
          .create()
          .add(Module2)
          .build();
      `))
        .rejects
        .toMatch('Property \'userService\' is missing in type \'{}\' but required in type \'Readonly<Pick<Module1, "userService">>\'.');
    });

    it('The build container should have the correct service declarations', async () => {
      // language=TypeScript
      return expect(declarationOfTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class Module1 implements Module {
          public halloService() {
            return { hallo: () => 'hallo', };
          }
        }

        export const container = ContainerFactory
          .create()
          .add(Module1);
      `))
        .resolves
        .toMatchSnapshot();
    });

    it('A module cannot have the same service name', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class Module1 implements Module {
          public halloService() {
          }
        }

        class Module2 implements Module {
          public halloService() {
          }
        }

        ContainerFactory
          .create()
          .add(Module1)
          .add(Module2);
      `))
        .rejects
        .toMatch(/Types of property 'halloService' are incompatible./);
    });

    it('registries function should be excluded from the container', async () => {
      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class Module1 implements Module {
          public registries() {
            return {
              personListeners: () => {
                return [];
              },
            };
          }
        }

        ContainerFactory
          .create()
          .add(Module1)
          .build().then(container => container.registries());
      `))
        .rejects
        .toMatch(/Property 'registries' does not exist on type./);
    });

    it('Register types should always match', async () => {

      // language=TypeScript
      return expect(compileTs(__dirname, `
        import { ContainerFactory } from '../ContainerFactory';
        import { Module } from '../Module';

        class Module1 implements Module {
          public registries() {
            return {
              // Template every listener should match to.
              someRegister: (): number[] => {
                return [];
              },
            };
          }
        }

        class Module2 implements Module {
          public registries() {
            return {
              // Template every listener should match to.
              someRegister: (): string[] => {
                return [];
              },
            };
          }
        }

        ContainerFactory
          .create()
          .add(Module1)
          .add(Module2)
          .build();
      `))
        .rejects
        .toMatch(/Types of property 'someRegister' are incompatible./);
    });

  });
});
