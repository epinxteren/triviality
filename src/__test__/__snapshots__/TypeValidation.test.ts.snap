// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeValidation ContainerFactory The build container should have the correct service declarations 1`] = `
"import { ContainerFactory } from '../ContainerFactory';
import { Module } from '../Module';
declare class Module1 implements Module {
    halloService(): {
        hallo: () => string;
    };
}
export declare const container: ContainerFactory<Pick<Module1, \\"halloService\\">, {}>;
export {};
"
`;
